File: 0278-first-bad-version/0278-first-bad-version.java
/* The isBadVersion API is defined in the parent class VersionControl.
   boolean isBadVersion(int version); */

public class Solution extends VersionControl {
  public int firstBadVersion(int n) {
    int l = 1;
    int r = n;
    if (l == r) return l;
    int mid = l + (r - l) / 2;
    while (l < r) {
      if (isBadVersion(mid)) {
        r = mid;
      } else {
        l = mid + 1;
      }
      mid = l + (r - l) / 2;
      System.out.println(mid);
    }
    return mid;
  }
}

File: 0278-first-bad-version/0278-first-bad-version.ts
/**
 * The knows API is defined in the parent class Relation.
 * isBadVersion(version: number): boolean {
 *   ...
 * };
 */

var solution = function(isBadVersion: any) {

  return function(n: number): number {
    let l = 1
    let r = n
    if (l === r) return l
    let mid = Math.floor((r + l) / 2)
    while (l < r) {
      if (isBadVersion(mid)) {
        r = mid
      } else {
        l = mid + 1
      }
      mid = Math.floor((r + l) / 2)
    }
    return mid
  };
};